File: 77486_다단계칫솔판매.py
class Person:
    def __init__(self, referral):
        self.money = 0
        self.referral = referral
    
    def Distribute(self, profit):
        if self.referral == None:
            self.money += profit
            return
        referralProfit = int(profit * 0.1)
        myProfit = profit - referralProfit
        self.money += myProfit
        if referralProfit > 0:
            self.referral.Distribute(referralProfit)

center = Person(None)
personDict = {'-':center}

def makeTree(enroll, referral):
    for (personName,referralName) in zip(enroll, referral):
        personDict[personName] = Person(personDict[referralName])
    return

def calculateMoney(seller, amount):
    for (s, a) in zip(seller, amount):
        personDict[s].Distribute(a*100)
    return

def getResult(enroll):
    result = [personDict[myName].money for myName in enroll]
    return result

def solution(enroll, referral, seller, amount):
    makeTree(enroll, referral)
    calculateMoney(seller, amount)
    answer = getResult(enroll)
    return answer

File: Programmers/77486_다단계칫솔판매.cpp
#include <string>
#include <vector>
#include <string>
#include <vector>
#include <map>

using namespace std;

class Person
{
    Person* parent;
    int money;
public:
    Person(Person* parent) : parent(parent)
    {
        money = 0;
    }
    int GetMoney() { return money; }
    void DistributeMoney(int profitMoney)
    {
        if (parent == NULL)
        {
            money += profitMoney;
            return;
        }

        int parenProfit = profitMoney * 0.1;
        int myProfit = profitMoney - parenProfit;
        money += myProfit;
        if (parenProfit > 0 && parent != NULL)
        {
            parent->DistributeMoney(parenProfit);
        } 
    }
};

map<string, Person*> personMap;

void makeTree(vector<string> enroll, vector<string> referral)
{
    personMap["-"] = new Person(NULL);
    for (int i=0;i<enroll.size();i++)
        personMap[enroll[i]] = new Person(personMap[referral[i]]);
}

void calMoney(vector<string> seller, vector<int> amount)
{
    for (int i = 0; i < seller.size(); i++)
        personMap[seller[i]]->DistributeMoney(amount[i]*100);
}

vector<int> getResult(vector<string> enroll)
{
    vector<int> result;
    for (int i = 0; i < enroll.size(); i++)
        result.push_back(personMap[enroll[i]]->GetMoney());
    return result;
}

void deleteTree()
{
    map<string, Person*>::iterator iter;
    for (iter = personMap.begin(); iter != personMap.end(); iter++)
    {
        delete iter->second;
    }
}

vector<int> solution(vector<string> enroll, vector<string> referral, vector<string> seller, vector<int> amount) {
    makeTree(enroll, referral);
    calMoney(seller, amount);
    vector<int> answer = getResult(enroll);
    deleteTree();
    return answer;
}

int main()
{
    vector<string> enroll = {"john", "mary", "edward", "sam", "emily", "jaimie", "tod", "young"};
    vector<string> referral = {"-", "-", "mary", "edward", "mary", "mary", "jaimie", "edward"};
    vector<string> seller = {"young", "john", "tod", "emily", "mary"};
    vector<int> amount = {12, 4, 2, 5, 10};
    vector<int> result = solution(enroll, referral, seller, amount);
    return 0;
}